File: 115-distinctSequence.py
class Solution:
    # @param {string} s
    # @param {string} t
    # @return {integer}
    def numDistinct(self, s, t):
        slen = len(s); tlen = len(t)
        dp = [[0 for j in xrange(tlen + 1)] for i in xrange(slen + 1)]
        for i in xrange(slen + 1):
            dp[i][0] = 1

        # self.print_matrix(dp)
        
        for i in xrange(1, slen + 1):
            for j in xrange(1, tlen + 1):
                if s[i - 1] == t[j - 1]:
                   dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j] 
                else:
                   dp[i][j] = dp[i - 1][j]

        # self.print_matrix(dp)

        return dp[slen][tlen]
    
    def print_matrix(self, matrix):
        for i in xrange(len(matrix)):
            print matrix[i]

if __name__ == '__main__':
   s = Solution()
   print s.numDistinct("rabbbit", "rabbit")

File: C++/115-distinctSequence.cpp
#include "include/common.h"

class Solution {
public:
    int numDistinct(string s, string t) {
      // Utils u;
      int tlen = t.length() + 1, slen = s.length() + 1;
      if (tlen == 1) {
        return 0;
      };
      vector<vector<int> > stTable(slen, vector<int>(tlen, 0));
      for (int i = 0; i < slen; i++) {
        stTable[i][0] = 1;
      };
      for (int j = 1; j < tlen; j++) {
        stTable[0][j] = 0;
      };
      for (int i = 1; i < slen; i++) {
        for (int j = 1; j < tlen; j++) {
          stTable[i][j] = stTable[i - 1][j] + (s[i - 1] == t[j - 1]) * stTable[i - 1][j - 1];
        };
      };
      // u.print2DIntVector(stTable);
      return stTable[slen - 1][tlen - 1];
    }
};

int main(void) {
  Solution s;
  cout << s.numDistinct("rabbbit", "rabit") << endl;
  return 0;
};